File: cpp/095_UniqueBinarySearchTreesII.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        if( n < 1 ) return vector<TreeNode*>();
        return recursive(1,n);
    }
    
private:
    vector<TreeNode *> recursive(int begin, int end){
        vector<TreeNode *> ret;
        if(begin > end){
            ret.push_back(NULL);
            return ret;
        }
        for(int i = begin; i <= end; ++i){
            auto left = recursive(begin,i-1);
            auto right = recursive(i+1,end);
            for(auto m : left){
                for(auto n : right){
                    auto node = new TreeNode(i);
                    node->left = m;
                    node->right = n;
                    ret.push_back(node);
                }
            }
        }
        return ret;
    }
};

File: js/095_UniqueBinarySearchTreesII.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number} n
 * @return {TreeNode[]}
 */
var generateTrees = function(n) {
    if(n < 1) return [];
    return recursive(1,n);
};

var recursive = function(b, e){
    var ret = [];
    if(b > e){
        ret.push(null);
        return ret;
    }
    for(var i = b; i <= e; ++i){
        var left = recursive(b, i-1);
        var right = recursive(i+1, e);
        for(var l in left){
            for(var r in right){
                var node = new TreeNode(i);
                node.left = left[l];
                node.right = right[r];
                ret.push(node);
            }
        }
    }
    return ret;
};